File: Habitica/src/main/java/com/habitrpg/android/habitica/models/tasks/Days.java
package com.habitrpg.android.habitica.models.tasks;

import android.os.Parcel;
import android.os.Parcelable;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class Days extends RealmObject implements Parcelable {

    @PrimaryKey
    private String taskId;
    private boolean m, t, w, th, f, s, su;

    public boolean getT() {
        return t;
    }

    public void setT(boolean t) {
        this.t = t;
    }

    public boolean getW() {
        return w;
    }

    public void setW(boolean w) {
        this.w = w;
    }

    public boolean getTh() {
        return th;
    }

    public void setTh(boolean th) {
        this.th = th;
    }

    public boolean getF() {
        return f;
    }

    public void setF(boolean f) {
        this.f = f;
    }

    public boolean getS() {
        return s;
    }

    public void setS(boolean s) {
        this.s = s;
    }

    public boolean getSu() {
        return su;
    }

    public void setSu(boolean su) {
        this.su = su;
    }

    public boolean getM() {
        return m;
    }

    public void setM(boolean m) {
        this.m = m;
    }

    public boolean getForDay(int day) {
        switch (day) {
            case 2:
                return this.getM();
            case 3:
                return this.getT();
            case 4:
                return this.getW();
            case 5:
                return this.getTh();
            case 6:
                return this.getF();
            case 7:
                return this.getS();
            case 1:
                return this.getSu();
        }
        return false;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.taskId);
        dest.writeByte(this.m ? (byte) 1 : (byte) 0);
        dest.writeByte(this.t ? (byte) 1 : (byte) 0);
        dest.writeByte(this.w ? (byte) 1 : (byte) 0);
        dest.writeByte(this.th ? (byte) 1 : (byte) 0);
        dest.writeByte(this.f ? (byte) 1 : (byte) 0);
        dest.writeByte(this.s ? (byte) 1 : (byte) 0);
        dest.writeByte(this.su ? (byte) 1 : (byte) 0);
    }

    protected Days(Parcel in) {
        this.taskId = in.readString();
        this.m = in.readByte() != 0;
        this.t = in.readByte() != 0;
        this.w = in.readByte() != 0;
        this.th = in.readByte() != 0;
        this.f = in.readByte() != 0;
        this.s = in.readByte() != 0;
        this.su = in.readByte() != 0;
    }

    public Days() {
        m = true;
        t = true;
        w = true;
        th = true;
        f = true;
        s = true;
        su = true;
    }

    public static final Parcelable.Creator<Days> CREATOR = new Parcelable.Creator<Days>() {
        @Override
        public Days createFromParcel(Parcel source) {
            return new Days(source);
        }

        @Override
        public Days[] newArray(int size) {
            return new Days[size];
        }
    };

    public String getTaskId() {
        return taskId;
    }

    public void setTaskId(String taskId) {
        this.taskId = taskId;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/models/tasks/Days.kt
package com.habitrpg.android.habitica.models.tasks

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.habitrpg.android.habitica.R

@io.realm.annotations.RealmClass(embedded = true)
open class Days() : io.realm.RealmObject(), Parcelable {
    val isEveryDay: Boolean
        get() = m && t && w && th && f && s && su
    val isOnlyWeekdays: Boolean
        get() = m && t && w && th && f && !s && !su
    val isOnlyWeekends: Boolean
        get() = !m && !t && !w && !th && !f && s && su
    var m: Boolean = true
    var t: Boolean = true
    var w: Boolean = true
    var th: Boolean = true
    var f: Boolean = true
    var s: Boolean = true
    var su: Boolean = true

    override fun writeToParcel(
        dest: Parcel,
        flags: Int,
    ) {
        dest.writeByte(if (m) 1.toByte() else 0.toByte())
        dest.writeByte(if (t) 1.toByte() else 0.toByte())
        dest.writeByte(if (w) 1.toByte() else 0.toByte())
        dest.writeByte(if (th) 1.toByte() else 0.toByte())
        dest.writeByte(if (f) 1.toByte() else 0.toByte())
        dest.writeByte(if (s) 1.toByte() else 0.toByte())
        dest.writeByte(if (su) 1.toByte() else 0.toByte())
    }

    protected constructor(`in`: Parcel) : this() {
        m = `in`.readByte().toInt() != 0
        t = `in`.readByte().toInt() != 0
        w = `in`.readByte().toInt() != 0
        th = `in`.readByte().toInt() != 0
        f = `in`.readByte().toInt() != 0
        s = `in`.readByte().toInt() != 0
        su = `in`.readByte().toInt() != 0
    }

    override fun describeContents(): Int {
        return 0
    }

    fun dayStrings(context: Context): List<String> {
        val days = mutableListOf<String>()
        if (m) days.add(context.getString(R.string.monday))
        if (t) days.add(context.getString(R.string.tuesday))
        if (w) days.add(context.getString(R.string.wednesday))
        if (th) days.add(context.getString(R.string.thursday))
        if (f) days.add(context.getString(R.string.friday))
        if (s) days.add(context.getString(R.string.saturday))
        if (su) days.add(context.getString(R.string.sunday))
        return days
    }

    companion object CREATOR : Parcelable.Creator<Days> {
        override fun createFromParcel(parcel: Parcel): Days {
            return Days(parcel)
        }

        override fun newArray(size: Int): Array<Days?> {
            return arrayOfNulls(size)
        }
    }
}
